vlan and bond support in networkd

sshedi opened the pull request:
<!--
Thank you for submitting a PR to cloud-init!

To ease the process of reviewing your PR, do make sure to complete the following checklist **before** submitting a pull request.

- [x] I have signed the CLA: https://ubuntu.com/legal/contributors
- [x] I have included a comprehensive commit message using the guide below
- [x] I have added unit tests to cover the new behavior under ``tests/unittests/``
  - Test files should map to source files i.e. a source file ``cloudinit/example.py`` should be tested by ``tests/unittests/test_example.py``
  - Run unit tests with ``tox -e py3``
- [ ] I have kept the change small, avoiding unnecessary whitespace or non-functional changes.
- [ ] I have added a reference to issues that this PR relates to in the PR message (Refs GH-1234, Fixes GH-1234)
- [ ] I have updated the documentation with the changed behavior.
  - If the change doesn't change the user interface and is trivial, this step may be skipped.
  - Cloud-config documentation is generated from the jsonschema.
  - Generate docs with ``tox -e docs``.
-->


## Proposed Commit Message
feat(networkd): support vlan and bond rendering

```
Networkd renderer used in Photon OS doesn't support vlan and bond settings specified through network config yaml.
This is an effort towards supporting the same in networkd renderer.
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->


## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

sshedi commented:
@TheRealFalcon @holmanb Can you PTAL at this PR? I'm not sure if the failure is legit. If not, please take this fix if things look good. Thanks.

TheRealFalcon commented on tests/unittests/net/artifacts/photon_net_config_v2.yaml:
```diff
@@ -0,0 +1,82 @@
+#cloud-config
```
This `#cloud-config` header doesn't belong

TheRealFalcon commented on tests/unittests/net/artifacts/photon_net_config_v2.yaml:
```diff
@@ -0,0 +1,82 @@
+        - 192.168.100.10/24
+        - 192.168.100.11/24
+      dhcp6: true
+      mtu: 1500
```
`mtu: 1500` is repeated in two different places in this file. Can we use a different value for each to ensure the right value is going to the right file?

TheRealFalcon commented on cloudinit/net/networkd.py:
```diff
@@ -46,33 +48,57 @@ def update_route_section(self, sec, rid, key, val):
                 )
                 self.conf_dict[k][rid].sort()
 
+    def normalize_dict(self, data):
```
The recursive nature of this introduces some typing issues. `data` can be a dict or list, which means `normalize_dict()` can return a dict or list. I think it might be best to introduce a helper function while having the top level function take `Dict[str, List]` and return `Dict[str, List]`.

Also, the original declaration of `self.conf_dict` no longer needs to be an OrderedDict now, correct?

TheRealFalcon commented on cloudinit/net/networkd.py:
```diff
@@ -46,33 +48,57 @@ def update_route_section(self, sec, rid, key, val):
     def get_final_conf(self):
         contents = ""
+
+        self.conf_dict = self.normalize_dict(self.conf_dict)
```
Nit: Is there a reason/need to reassign `self.conf_dict` here?

TheRealFalcon commented on cloudinit/net/networkd.py:
```diff
@@ -296,12 +325,36 @@ def render_network_state(
+        vlan_link = {}
+        bond_link = {}
+
+        if ns.version == 2:
```
Does this mean only support vlans and bonds if using configuration v2? If so, why only v2?

sshedi commented:
> @sshedi , I left some comments inline, but overall things look good!

Thank you so much for the review and inputs. Tried addressing all your suggestions. PTAL when you get time.
